### ISSUE TITLE
Ion-Virtual-Scroll or Cdk-Virtual-Scroll?

### ISSUE THREAD

**Issue by GosuMania** [ionic-team/ionic-framework#23359]
I am trying to use **cdk-virtual-scroll** instead of **ion-virtual-scroll** because with ion I have problems with overlapping items when they have different heights.

But cdk introduces several problems:
1. I can't use ion-refresher properly, when I scroll up the list the ion-refresher is continuously called up even if it is placed outside the cdk
2. I can't add an ion-slides to the top of the list, it looks awful and doesn't work well
3. the scrolling of the list in ascent is not fluid as well as in descent

These are just some of the problems that I didn't have with ion-virtual-scroll, but I had the terrible problem that there is no autorisize!

Possible there is no solution? Should I also use ion-list with a list of 1000 items? The problem would then be that if I change the page and return to the list and want it to return to the exact position in which it was left, the application will be very slow because it will load as many items as it had previously loaded but a little at a time .. ..

**Comment by elvisgraho** [comment]
I use ngx-virtual-scroller in my project and have no issues. 

``` html
<ion-list lines="none">
  <ion-refresher
    slot="fixed"
    (ionRefresh)="doRefresh($event)"
  >
    <ion-refresher-content
      refreshingSpinner="circles"
    >
    </ion-refresher-content>
  </ion-refresher>

    <virtual-scroller
      [parentScroll]="ionContent"
      #scroll
      [items]="protocols"
    >
      <the-item
        *ngFor="let protocol of scroll.viewPortItems"
      ></the-item>
    </virtual-scroller>

</ion-list>

```

**Comment by GosuMania** [comment]
> I use ngx-virtual-scroller in my project and have no issues.
> 
> ```
> <ion-list lines="none">
>   <ion-refresher
>     slot="fixed"
>     (ionRefresh)="doRefresh($event)"
>   >
>     <ion-refresher-content
>       refreshingSpinner="circles"
>     >
>     </ion-refresher-content>
>   </ion-refresher>
> 
>     <virtual-scroller
>       [parentScroll]="ionContent"
>       #scroll
>       [items]="protocols"
>     >
>       <the-item
>         *ngFor="let protocol of scroll.viewPortItems"
>       ></the-item>
>     </virtual-scroller>
> 
> </ion-list>
> ```
> 
> ```ts
>   doRefresh(refresher) {
>     this.refresher = refresher;
>     this.asyncAction();
>   }
>   
>   // when completed
>   this.refresher && this.refresher.detail.complete();
> ```
> 
> For ionContent. I have it in one central component, thus i need to save it in a service so I can share it across components. In my central component (that has tag in html) I do this
> 
> ```ts
>   @ViewChild(IonContent, { static: false }) ionContent: IonContent;
> 
>   ngAfterViewInit() {
>     this.ionContent.getScrollElement().then((scrollElement) => {
>       this.utilsService.setIonContent(scrollElement);
>     });
>   }
> ```

i try add this component with `ng add ngx-virtual-scroller@3.0.3.`

but i have this error:
`The package that you are trying to add does not support schematics. You can try using a different version of the package or contact the package author to add ng-add support.`

**Comment by elvisgraho** [comment]
" The problem would then be that if I change the page and return to the list and want it to return to the exact position in which it was left"

I use service on that one where i track scroll position on router changes and then use ionContent.scrollToPoint(0, scrollToVal, 500); when I need.

**Comment by elvisgraho** [comment]
I assume you use Angular. 
``` sh
npm i ngx-virtual-scroller
```

**Comment by GosuMania** [comment]
> ```shell
> ```shell
> npm i ngx-virtual-scroller
> ```
> 
> 
>     
>       
>     
> 
>       
>     
> 
>     
>   
>       
> ```

yes i try it before but i have this error: 

`npm ERR! code ERESOLVE
npm ERR! ERESOLVE unable to resolve dependency tree
npm ERR! 
npm ERR! While resolving: gosumania@0.0.1
npm ERR! Found: tslib@2.2.0
npm ERR! node_modules/tslib
npm ERR!   tslib@"^2.0.0" from the root project
npm ERR!   tslib@"^2.0.0" from @angular/core@11.2.14
npm ERR!   node_modules/@angular/core
npm ERR!     @angular/core@"~11.2.7" from the root project
npm ERR!     peer @angular/core@">=6.0.0" from ngx-virtual-scroller@4.0.3
npm ERR!     node_modules/ngx-virtual-scroller
npm ERR!       ngx-virtual-scroller@"*" from the root project
npm ERR! 
npm ERR! Could not resolve dependency:
npm ERR! peer tslib@"^1.10.0" from ngx-virtual-scroller@4.0.3
npm ERR! node_modules/ngx-virtual-scroller
npm ERR!   ngx-virtual-scroller@"*" from the root project
npm ERR! 
npm ERR! Fix the upstream dependency conflict, or retry
npm ERR! this command with --force, or --legacy-peer-deps
npm ERR! to accept an incorrect (and potentially broken) dependency resolution.
npm ERR! 
npm ERR! See /Users/gosumania/.npm/eresolve-report.txt for a full report.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/gosumania/.npm/_logs/2021-05-25T12_44_30_137Z-debug.log`

**Comment by elvisgraho** [comment]
I provided an npm package as an alternative to ion-virtual-scroll that works for me and fixes issues that I had with ion-virtual-scroll. That should be enough.

**Comment by GosuMania** [comment]
> I provided an npm package as an alternative to ion-virtual-scroll that works for me and fixes issues that I had with ion-virtual-scroll. That should be enough. Here is the link - https://github.com/rintoj/ngx-virtual-scroller

Can you help me, I need to scroll all the content as I scroll, and not just the list in virtual-scroller, so the items and slides above automatically disappear as soon as I start going down the list.... Example of my HTML:
`
<ion-content>
      <ion-refresher slot="fixed" (ionRefresh)="doRefresh($event)" [disabled]="false">
        <ion-refresher-content pullingIcon="chevron-down-circle-outline">
        </ion-refresher-content>
      </ion-refresher>
      
     <ion-slides> example...</ion-slides>
     
     <ion-item>example...</ion-item>
     
     <ion-item>example...</ion-item>
     
     <virtual-scroller [parentScroll]="ionContent"  #scroll [items]="deals" *ngIf="deals.length > 0"> 
          <ion-item   *ngFor="let deal of scroll.viewPortItems" class="ion-no-padding">
               
          </ion-item>
     </virtual-scroller>
     
</ion-content>
  
  `

**Comment by elvisgraho** [comment]
> > I provided an npm package as an alternative to ion-virtual-scroll that works for me and fixes issues that I had with ion-virtual-scroll. That should be enough. Here is the link - https://github.com/rintoj/ngx-virtual-scroller
> 
> Can you help me, I need to scroll all the content as I scroll, and not just the list in virtual-scroller, so the items and slides above automatically disappear as soon as I start going down the list.... Example of my HTML:
> 
> ```
> `<ion-content>
>       <ion-refresher slot="fixed" (ionRefresh)="doRefresh($event)" [disabled]="false">
>         <ion-refresher-content pullingIcon="chevron-down-circle-outline">
>         </ion-refresher-content>
>       </ion-refresher>
>       
>      <ion-slides> example...</ion-slides>
>      
>      <ion-item>example...</ion-item>
>      
>      <ion-item>example...</ion-item>
>      
>      <virtual-scroller [parentScroll]="ionContent"  #scroll [items]="deals" *ngIf="deals.length > 0"> 
>           <ion-item   *ngFor="let deal of scroll.viewPortItems" class="ion-no-padding">
>                
>           </ion-item>
>      </virtual-scroller>
>      
> </ion-content> `
> ```

If you wrap everything besides ion-content in ion-list like I did in the first example will that change things?

**Comment by liamdebeasi** [comment]
Thanks for the issue. Can you provide a code reproduction for the issues with Ionic components when using CDK Virtual Scroll?

**Comment by liamdebeasi** [comment]
Hey there,

I did some more digging into this. We are aware that some Ionic components (like `ion-refresher`) do not work great in a virtual scrolling environment. I created https://github.com/ionic-team/ionic-framework/issues/23437 which we will use to improve our compatibility between these components and virtual scrollers. Starting with Framework v6 we will be recommending CDK Scroller over `ion-virtual-scroll`. We will also be releasing new documentation that better details how to correctly use virtual scrollers in Ionic Framework. 

`ion-slides` is powered by Swiper, and I recommend trying the virtual scrolling feature it provides. You will need to use the library directly as `ion-slides` does not have that feature bundled with it: https://swiperjs.com/

For list scrolling performance in 3rd party virtual scrolling solutions, that mainly depends on the virtual scroll library's performance. In this case, we recommend checking the virtual scroller's documentation as it may have some tips for optimizing scrolling performance. If that still does not fix the issue, we recommend creating a minimal code reproduction and filing an issue with the virtual scroll maintainer.

I am going to close this as we will track compatibility progress in https://github.com/ionic-team/ionic-framework/issues/23437. Improved documentation will be rolled out with Framework v6 (coming soon!). Thanks!
